subsys/zigbee/cli/zigbee_cli_cmd_zcl_groups.c

```diff
@@ -0,0 +1,996 @@
+ *
+ * @param index   Index of the entry with frame to send in the context manager.
+ */
+static void zb_zcl_send_add_group_if_identifying_cmd(zb_uint8_t index)
+{
+	struct zcl_packet_info *packet_info;
+	struct ctx_entry *entry = ctx_mgr_get_entry_by_index(index);
+
+	if (entry == NULL) {
+		LOG_ERR("Couldn't send the ZCL command frame - context entry %d not found", index);
+		return;
+	}
+
+	/* Get the ZCL packet sequence number. */
+	entry->id = ZCL_CTX().seq_number;
+
+	packet_info = &(entry->groups_cmd_data.packet_info);
+
+	ZB_ZCL_GROUPS_SEND_ADD_GROUP_IF_IDENT_REQ(
+				packet_info->buffer,
+				packet_info->dst_addr,
+				packet_info->dst_addr_mode,
+				packet_info->dst_ep,
+				packet_info->ep,
+				packet_info->prof_id,
+				ZB_ZCL_DISABLE_DEFAULT_RESPONSE,
+				packet_info->cb,
+				entry->groups_cmd_data.group_id);
+}
+
+/**@brief Function to construct and send Remove All Groups command.
+ *
+ * @param index   Index of the entry with frame to send in the context manager.
+ */
+static void zb_zcl_send_remove_all_groups_cmd(zb_uint8_t index)
+{
+	struct zcl_packet_info *packet_info;
+	struct ctx_entry *entry = ctx_mgr_get_entry_by_index(index);
+
+	if (entry == NULL) {
+		LOG_ERR("Couldn't send the ZCL command frame - context entry %d not found", index);
+		return;
+	}
+
+	/* Get the ZCL packet sequence number. */
+	entry->id = ZCL_CTX().seq_number;
+
+	packet_info = &(entry->groups_cmd_data.packet_info);
+
+	ZB_ZCL_GROUPS_SEND_REMOVE_ALL_GROUPS_REQ(
+				packet_info->buffer,
+				packet_info->dst_addr,
+				packet_info->dst_addr_mode,
+				packet_info->dst_ep,
+				packet_info->ep,
+				packet_info->prof_id,
+				ZB_ZCL_DISABLE_DEFAULT_RESPONSE,
+				packet_info->cb);
+}
+
+/**@brief Handles timeout error and invalidates group command entry in context manager.
+ *
+ * @param[in] index   Index of context entry with group command data to invalidate.
+ */
+static void zcl_groups_cmd_timeout(zb_uint8_t index)
+{
+	struct ctx_entry *entry = ctx_mgr_get_entry_by_index(index);
+
+	if (entry == NULL) {
+		LOG_ERR("Couldn't get entry %d - entry not found", index);
+		return;
+	}
+
+	zb_cli_print_error(entry->shell, "Request timed out", ZB_FALSE);
+
+	ctx_mgr_delete_entry(entry);
+}
+
+/**@brief Function to send the ZCL groups cmds.
+ *
+ * @param entry   Pointer to the entry with frame to send in the context manager.
+ *
+ * @return Error code with function execution status.
+ */
+static int zcl_groups_cmd_send(struct ctx_entry *entry)
+{
+	zb_ret_t zb_err_code;
+	uint8_t entry_index = ctx_mgr_get_index_by_entry(entry);
+
+	zb_err_code = ZB_SCHEDULE_APP_ALARM(zcl_groups_cmd_timeout, entry_index,
+					    (CMD_ENTRY_TIMEOUT_S * ZB_TIME_ONE_SECOND));
+	if (zb_err_code != RET_OK) {
+		zb_cli_print_error(entry->shell, "Couldn't schedule timeout cb.", ZB_FALSE);
+		goto cmd_send_error;
+	}
+
+	zb_err_code = ZB_SCHEDULE_APP_CALLBACK(entry->groups_cmd_data.send_fn, entry_index);
```

Comment:
That's an interesting idea!
Maybe we should think about add the "callback with timeout" API in the future?

Reply:
I will add this to the "Refactor Zigbee shell" bucket list